pkg/kubectl/cmd/log.go

```diff
@@ -72,115 +72,144 @@ func selectContainer(pod *api.Pod, in io.Reader, out io.Writer) string {
-	cmd.Flags().Int("limit-bytes", 0, "Maximum bytes of logs to return. Defaults to no limit.")
-	cmd.Flags().Int("tail", -1, "Lines of recent log file to display. Defaults to -1, showing all log lines.")
-	cmd.Flags().String("since-time", "", "Only return logs after a specific date (RFC3339). Defaults to all logs. Only one of since-time / since may be used.")
-	cmd.Flags().Duration("since", 0, "Only return logs newer than a relative duration like 5s, 2m, or 3h. Defaults to all logs. Only one of since-time / since may be used.")
-	cmd.Flags().StringVarP(&params.containerName, "container", "c", "", "Container name")
+	cmd.Flags().BoolVarP(&o.Follow, "follow", "f", o.Follow, "Specify if the logs should be streamed.")
+	cmd.Flags().BoolVar(&o.Timestamps, "timestamps", o.Timestamps, "Include timestamps on each line in the log output")
+	cmd.Flags().BoolVar(&o.Interactive, "interactive", o.Interactive, "If true, prompt the user for input when required. Default true.")
+	cmd.Flags().IntVar(&o.LimitBytes, "limit-bytes", o.LimitBytes, "Maximum bytes of logs to return. Defaults to no limit.")
+	cmd.Flags().BoolVarP(&o.Previous, "previous", "p", o.Previous, "If true, print the logs for the previous instance of the container in a pod if it exists.")
+	cmd.Flags().IntVar(&o.Tail, "tail", o.Tail, "Lines of recent log file to display. Defaults to -1, showing all log lines.")
+	cmd.Flags().StringVar(&o.SinceTime, "since-time", o.SinceTime, "Only return logs after a specific date (RFC3339). Defaults to all logs. Only one of since-time / since may be used.")
+	cmd.Flags().DurationVar(&o.SinceSeconds, "since", o.SinceSeconds, "Only return logs newer than a relative duration like 5s, 2m, or 3h. Defaults to all logs. Only one of since-time / since may be used.")
+	cmd.Flags().StringVarP(&o.ContainerName, "container", "c", o.ContainerName, "Container name")
 	return cmd
 }
 
-// RunLog retrieves a pod log
-func RunLog(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, p *logParams) error {
-	if len(os.Args) > 1 && os.Args[1] == "log" {
-		printDeprecationWarning("logs", "log")
-	}
-
-	if len(args) == 0 {
+func (o *LogsOptions) Complete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string) error {
+	switch len(args) {
+	case 0:
 		return cmdutil.UsageError(cmd, "POD is required for log")
-	}
 
-	if len(args) > 2 {
-		return cmdutil.UsageError(cmd, "log POD [CONTAINER]")
-	}
+	case 1:
+		o.PodName = args[0]
+	case 2:
+		o.PodName = args[0]
+		o.ContainerName = args[1]
 
-	sinceSeconds := cmdutil.GetFlagDuration(cmd, "since")
-	sinceTime := cmdutil.GetFlagString(cmd, "since-time")
-	if len(sinceTime) > 0 && sinceSeconds > 0 {
-		return cmdutil.UsageError(cmd, "only one of --since, --since-time may be specified")
+	default:
+		return cmdutil.UsageError(cmd, "log POD [CONTAINER]")
 	}
 
-	namespace, _, err := f.DefaultNamespace()
+	var err error
+	o.PodNamespace, _, err = f.DefaultNamespace()
```

Comment:
> inline the err != nil (here and below)

```go
if o.PodNamespace, _, err := f.DefaultNamespace(); err != nil {
```
doesn't compile because `o.PodNamespace` already exists.